lib/pbench/test/unit/server/test_unpack_tarballs.py

```diff
@@ -0,0 +1,293 @@
+        obj = UnpackTarballs(server_config, make_logger)
+        result = obj.unpack_tarballs(self.min_size, self.max_size)
+        for tar in self.tarlist:
+            assert (
+                f"Unpacking of tarball {tar} failed: An error occurred while unpacking tarball: tar exited with status 1"
+                in caplog.text
+            )
+        assert result.total == 2
+        assert result.success == 0
+
+    def test_unpack_tarballs_unlink_failure(
+        self, monkeypatch, server_config, make_logger, caplog
+    ):
+        """Show that when unlinking failure leads to FileNotFoundError,
+        it is handled successfully."""
+
+        def mock_unlink(path):
+            """
+            Mock of os.unlink() function
+
+            Args:
+                path: tarball path in 'TO-INDEX' state directory
+
+            Returns:
+                Raises FileNotFoundError Exception
+            """
+            raise FileNotFoundError(f"No such file or directory: '{path}'")
+
+        monkeypatch.setattr(Path, "glob", lambda path, args: self.tarlist)
+        monkeypatch.setattr(Path, "resolve", lambda self, strict: self)
+        monkeypatch.setattr(Path, "stat", self.mock_stat)
+        monkeypatch.setattr(
+            "pbench.server.unpack_tarballs.get_tarball_md5", lambda tb: "my_md5_string"
+        )
+        monkeypatch.setattr(FileTree, "unpack", lambda self, dataset_id: None)
+        monkeypatch.setattr(os, "symlink", lambda path, dest: None)
+        monkeypatch.setattr(os, "unlink", mock_unlink)
+        obj = UnpackTarballs(server_config, make_logger)
+        result = obj.unpack_tarballs(self.min_size, self.max_size)
+        for tar in self.tarlist:
+            assert (
+                f"Error in removing symlink from TO-UNPACK state. No such file or directory: '{server_config.ARCHIVE / 'ABC' / obj.LINKSRC / Path(tar).name}'"
+                in caplog.text
+            )
+        assert result.total == 2
+        assert result.success == 0
+
+    def test_unpack_tarballs_success(
+        self, monkeypatch, server_config, make_logger, caplog
+    ):
+        """Show that when unpacking goes smoothly we can count the
+        successfully unpacked tarballs."""
+
+        monkeypatch.setattr(Path, "glob", lambda path, args: self.tarlist)
+        monkeypatch.setattr(Path, "resolve", lambda self, strict: self)
+        monkeypatch.setattr(Path, "stat", self.mock_stat)
+        monkeypatch.setattr(
+            "pbench.server.unpack_tarballs.get_tarball_md5", lambda tb: "my_md5_string"
+        )
+        monkeypatch.setattr(FileTree, "unpack", lambda self, dataset_id: None)
+        monkeypatch.setattr(os, "symlink", lambda path, dest: None)
+        monkeypatch.setattr(os, "unlink", lambda path: None)
+        obj = UnpackTarballs(server_config, make_logger)
+        result = obj.unpack_tarballs(self.min_size, self.max_size)
+        assert result.total == 2
+        assert result.success == 2
```

Comment:
Consider adding code which checks that all the mocks were called as expected.  Ideally, they would check their inputs, but if we just want to "take attendance", something like this might suffice:
```
        ...
        mocks_called = []

        def tickmock(id: str, ret_val: Any) -> Any:
            mocks_called.append(id)
            return ret_val

        monkeypatch.setattr(
            Path, "glob", lambda path, args: tickmock("glob", self.tarlist)
        )
        monkeypatch.setattr(
            Path, "resolve", lambda self, strict: tickmock("resolve", self)
        )
        monkeypatch.setattr(
            Path,
            "stat",
            lambda self, *args, **kwargs: tickmock(
                "stat", TestUnpackTarballs.mock_stat(self, *args, **kwargs)
            ),
        )
        monkeypatch.setattr(
            "pbench.server.unpack_tarballs.get_tarball_md5",
            lambda tb: tickmock("get_tarball_md5", "my_md5_string"),
        )
        monkeypatch.setattr(
            FileTree, "unpack", lambda self, dataset_id: tickmock("unpack", None)
        )
        monkeypatch.setattr(os, "symlink", lambda path, dest: tickmock("symlink", None))
        monkeypatch.setattr(os, "unlink", lambda path: tickmock("unlink", None))
        obj = UnpackTarballs(server_config, make_logger)
        ...
        assert mocks_called == [
            "glob",
            *("stat",) * result.total,
            *("resolve", "unpack", "update_symlink") * result.success,
        ]
```
The `tickmock()` helper function allows us to do two things in a lambda expression where we would normally be able to do only one:  append to the list _and_ return a value.  Unfortunately, its addition makes Black wrap most of the lines.  🙁  (Disclaimer:  I haven't actually tried running this code; in particular, the `stat` mock may not be quite right....)

The `mocks_called` list comparison at the end will need to be corrected, depending on what you actually end up mocking.  The `*()` syntax decomposes the tuples, so their contents become part of the list (instead of using the tuples themselves as list elements); the other `*` causes the tuple-contents to be repeated the (hopefully) appropriate number of times.  (The comma at the end of the one-element tuple is required in order to designate it as a tuple -- without it, it's just a parenthesized expression, and the repetition operator is applied to each letter instead of the whole string!)